boa/src/syntax/lexer/string.rs

```diff
@@ -84,108 +82,81 @@ impl<R> Tokenizer<R> for StringLiteral {
                             "unterminated escape sequence in string literal",
                         ))
                     })?;
+
                     if escape != '\n' {
-                        let escaped_ch = match escape {
-                            'n' => '\n',
-                            'r' => '\r',
-                            't' => '\t',
-                            'b' => '\x08',
-                            'f' => '\x0c',
-                            '0' => '\0',
+                        match escape {
+                            'n' => buf.push('\n' as u16),
+                            'r' => buf.push('\r' as u16),
+                            't' => buf.push('\t' as u16),
+                            'b' => buf.push('\x08' as u16),
+                            'f' => buf.push('\x0c' as u16),
+                            '0' => buf.push('\0' as u16),
                             'x' => {
-                                let mut nums = [0u8; 2];
-                                cursor.fill_bytes(&mut nums)?;
-                                let nums = str::from_utf8(&nums).expect("non-UTF-8 bytes found");
-
-                                let as_num = match u64::from_str_radix(&nums, 16) {
-                                    Ok(v) => v,
-                                    Err(_) => 0,
-                                };
-                                match from_u32(as_num as u32) {
-                                    Some(v) => v,
-                                    None => {
-                                        return Err(Error::syntax(
-                                            format!(
-                                                "{}: {} is not a valid Unicode scalar value",
-                                                cursor.pos(),
-                                                as_num
-                                            ),
+                                let mut code_point_utf8_bytes = [0u8; 2];
+                                cursor.fill_bytes(&mut code_point_utf8_bytes)?;
+                                let code_point_str = str::from_utf8(&code_point_utf8_bytes)
+                                    .expect("malformed Hexadecimal character escape sequence");
```

Comment:
Shouldn't we return `Err`, This would panic if we gave it an invalid utf8 sequence, right?

Reply:
I found this panic being unreachable here. When the user input has invalid utf8 chars, boa always panics with InvalidData error "stream did not contain valid UTF-8", no matter the invalid char is in Unicode escape sequence, normal string char, identifiers, or even comments.

It seems to be a lexer-wide issue. Maybe we should have another issue/PR look into the behaviors?

Reply:
> It seems to be a lexer-wide issue. Maybe we should have another issue/PR look into the expected behaviors?

Yes. we can leave this for a separate issue/PR :)